### PR TITLE
Documentable components

### PR DESCRIPTION
**Pull request by tsujigiri**
See the changes to the README for the things this introduces.

### CODE REVIEW

File: spec/feature/styleguide_spec.rb [hunk 1]
```diff
@@ -0,0 +1,9 @@
+RSpec.feature 'Styleguide', :js do
+  scenario 'all components are properly documented' do
```
Comment (dziemba) [inline spec/feature/styleguide_spec.rb]: How exactly does this test ensure that `all components are properly documented`?

File: spec/feature/component_generator_spec.rb [hunk 2]
```diff
@@ -0,0 +1,23 @@
+RSpec.feature 'Component generator', :js do
+  let(:test_component_path) do
+    UiComponents::Engine.root.join('app', 'cells', 'test_component')
+  end
+
+  before do
+    expect(Dir.exist?(test_component_path)).to be(false)
+    `bundle exec thor generate_component test_component`
```
Comment (dziemba) [inline spec/feature/component_generator_spec.rb]: Shouldn't you check for errors here?

File: lib/ui_components/engine.rb [hunk 3]
```diff
@@ -1,4 +1,18 @@
...
+    initializer 'ui_components.assets.add_load_paths' do |app|
+      app.config.assets.paths += self.class.components_paths
+    end
+
+    initializer 'ui_components.assets.precompile' do |app|
+      app.config.assets.precompile += %w(styleguide.js
+                                         styleguide.css)
+    end
+
+    def self.components_paths
```
Comment (dziemba) [inline lib/ui_components/engine.rb]: Isn't this just a very long way of saying `Dir['app/cells/*/']`?

File: lib/ui_components/docu_cop.rb [hunk 4]
```diff
@@ -0,0 +1,77 @@
+module UiComponents
+  module DocuCop
```
Comment (dziemba) [inline lib/ui_components/docu_cop.rb]: If I understood this correctly, this class does two thing:
- check the documentation
- ensure valid parameters are given at runtime

Shouldn't that be two completely separate classes?

File: lib/generators/generate_component.thor [hunk 5]
```diff
@@ -0,0 +1,34 @@
+# vim: syntax=ruby
```
Comment (dziemba) [inline lib/generators/generate_component.thor]: Ew.

File: README.md [hunk 6]
```diff
@@ -1,7 +1,9 @@
 # UiComponents [![Circle CI](https://circleci.com/gh/ad2games/ui_components.svg?style=svg)](https://circleci.com/gh/ad2games/ui_components)
 
 This gem provides common UI components for all
-[ad2games](http://www.ad2games.com/) projects.
+[ad2games](http://www.ad2games.com/) projects as well as related documentation.
+To view the documentation run the dummy app within the spec directory and
```
Comment (dziemba) [inline README.md]: Can we make this somehow easier? Always typing `cd spec/dummy; rails s` feels wrong.

File: app/assets/stylesheets/ui_components.scss.erb [hunk 7]
```diff
@@ -7,3 +7,12 @@
 @import 'bootstrap-chosen';
 
 @import 'ui_components/*';
+
+<% UiComponents::Engine.components_paths.each do |path| %>
+  <%
```
Comment (dziemba) [inline app/assets/stylesheets/ui_components.scss.erb]: Don't we need `<%=` here?

File: app/cells/markdown_readonly/markdown_readonly.slim [hunk 8]
```diff
@@ -0,0 +1 @@
+div{ data-toggle="markdown-readonly" data-markdown=URI.escape(options[:content]) }
```
Comment (dziemba) [inline app/cells/markdown_readonly/markdown_readonly.slim]: `URI.escape`? Why not JSON?

File: app/cells/select/select_cell.rb [hunk 9]
```diff
@@ -0,0 +1,48 @@
+class SelectCell < FormCellBase
+  attribute :form, mandatory: true, description: 'A form object.'
```
Comment (dziemba) [inline app/cells/select/select_cell.rb]: Shouldn't attributes have some kind of type? How do I know if `classes` has to be an array or a space-separated string?

File: app/cells/markdown_readonly/markdown_readonly.coffee [hunk 10]
```diff
@@ -0,0 +1,7 @@
+#= require jquery
```
Comment (dziemba) [inline app/cells/markdown_readonly/markdown_readonly.coffee]: Don't we require all the libs globally anyway?

File: config/routes.rb [hunk 11]
```diff
@@ -0,0 +1,3 @@
+UiComponents::Engine.routes.draw do
```
Comment (dziemba) [inline config/routes.rb]: So you are supposed to mount the styleguide in the rails app that uses ui_c? Is there a nice way to do so? What about security?

File: lib/ui_components/cell.rb [hunk 12]
```diff
@@ -1,5 +1,14 @@
+require_relative '../../app/helpers/ui_components/view_helper'
```
Comment (dziemba) [inline lib/ui_components/cell.rb]: This doesn't feel right. Are you sure `ViewHelper` doesn't do too many things?

File: lib/ui_components/docu_cop.rb [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+      end.to_h
+    end
+
+    private
+
+    def validate_mandatory_attributes
+      missing_arguments = self.class.attributes
+        .select { |a| a[:mandatory] && public_send(a[:name]).nil? }
+      return unless missing_arguments.present?
+      fail MandatoryPropertyNotSet,
```
Comment (dziemba) [inline lib/ui_components/docu_cop.rb]: How exactly does this naming scheme work? We have `attributes`, `arguments` and `properties`?

File: lib/ui_components/docu_cop.rb [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+        options = ActionController::Parameters.new(options)
+        attributes << {
+          name: name,
+          mandatory: options[:mandatory] == true,
+          description: options.require(:description)
+        }
+        define_accessors(name)
+      end
+
+      def define_accessors(name)
```
Comment (dziemba) [inline lib/ui_components/docu_cop.rb]: Isn't this just what `attr_accessor` does?

File: ui_components.gemspec [hunk 15]
```diff
@@ -36,6 +36,7 @@ Gem::Specification.new do |s|
   s.add_dependency 'rails-assets-bootstrap-chosen'
   s.add_dependency 'rails-assets-marked'
   s.add_dependency 'rails-assets-ajax-chosen'
+  s.add_dependency 'rails-assets-highlightjs'
```
Comment (dziemba) [inline ui_components.gemspec]: Why is this a dependency of the gem but `bootstrap-sass` is only defined in the dummy app Gemfile?

File: app/cells/hello_world/hello_world_cell.rb [hunk 16]
```diff
@@ -0,0 +1,3 @@
+class HelloWorldCell < UiComponents::Cell
```
Comment (dziemba) [inline app/cells/hello_world/hello_world_cell.rb]: Do we still need this demo cell if we have actual components now?

File: spec/feature/styleguide_spec.rb (hunk 1 above)
Comment (tsujigiri) [inline spec/feature/styleguide_spec.rb]: Exceptions are thrown when the component doesn't have a description and at least one example.

File: spec/feature/component_generator_spec.rb (hunk 2 above)
Comment (tsujigiri) [inline spec/feature/component_generator_spec.rb]: I guess. If it fails the test fails, though.

File: lib/ui_components/docu_cop.rb (hunk 4 above)
Comment (tsujigiri) [inline lib/ui_components/docu_cop.rb]: It's also both defined over the `attribute` method. But I will check if I can separate this somewhat cleanly.

File: lib/generators/generate_component.thor (hunk 5 above)
Comment (tsujigiri) [inline lib/generators/generate_component.thor]: Yeah, I agree. I will adapt my vim config instead.

File: README.md (hunk 6 above)
Comment (tsujigiri) [inline README.md]: Sure.

File: spec/feature/styleguide_spec.rb (hunk 1 above)
Comment (dziemba) [inline spec/feature/styleguide_spec.rb]: Okay. Can you document that somehow? This just looks line an unfinished test.

File: app/assets/stylesheets/ui_components.scss.erb (hunk 7 above)
Comment (tsujigiri) [inline app/assets/stylesheets/ui_components.scss.erb]: Apparently not. I suppose `require_asset` works more procedurally than functional. I will check, though, if it still works in production mode... :D

File: app/cells/markdown_readonly/markdown_readonly.slim (hunk 8 above)
Comment (tsujigiri) [inline app/cells/markdown_readonly/markdown_readonly.slim]: Why not `URI.escape`?

File: app/cells/select/select_cell.rb (hunk 9 above)
Comment (tsujigiri) [inline app/cells/select/select_cell.rb]: We were considering this but wanted to proceed one step at a time. It's on the roadmap... ;)

File: app/assets/stylesheets/ui_components.scss.erb (hunk 7 above)
Comment (dziemba) [inline app/assets/stylesheets/ui_components.scss.erb]: Right, we don't. The documentation is just confusing. http://www.rubydoc.info/github/sstephenson/sprockets/Sprockets%2FContext%3Arequire_asset

Then why do we even need to do this in the scss file? Can't we move it to `engine.rb`?

File: app/cells/select/select_cell.rb (hunk 9 above)
Comment (dziemba) [inline app/cells/select/select_cell.rb]: Ok.

File: app/cells/markdown_readonly/markdown_readonly.slim (hunk 8 above)
Comment (dziemba) [inline app/cells/markdown_readonly/markdown_readonly.slim]: Because it's larger, slower and confusing to every developer who looks at it ;)

Comment (tsujigiri) [inline app/cells/markdown_readonly/markdown_readonly.slim]: As opposed to JSON, it works in this case, though. ;)

File: app/cells/markdown_readonly/markdown_readonly.coffee (hunk 10 above)
Comment (tsujigiri) [inline app/cells/markdown_readonly/markdown_readonly.coffee]: The idea here was to not require the dependencies globally anymore but have every component require their own dependencies. If there are any dependencies not required by any component, they are not included. That way we could even include only selected components and their dependencies in a project. In order to do this in a way that we don't just forget about including the right dependencies and having it break if the component actually is included individually, we would need a feature test for each component, only including that one component. This is another thing I would rather do separately.

tl;dr: I'll remove it for now. ;)

File: config/routes.rb (hunk 11 above)
Comment (tsujigiri) [inline config/routes.rb]: So far it is only done in the dummy app and I think that is enough, since we can now just run `rake styleguide` to run it via the dummy app.

File: lib/ui_components/cell.rb (hunk 12 above)
Comment (tsujigiri) [inline lib/ui_components/cell.rb]: The only thing it does is providing the interface to the outside world. Using this *from the inside* as well gives us the benefit of not having to change all the places where a component is used within another one, when the implementation changes and the interface stays the same.

File: lib/ui_components/docu_cop.rb (hunk 13 above)
Comment (tsujigiri) [inline lib/ui_components/docu_cop.rb]: I'm going to clean that up. We had to rename on the way since `property` is already used by the cells gem.

File: lib/ui_components/docu_cop.rb (hunk 14 above)
Comment (tsujigiri) [inline lib/ui_components/docu_cop.rb]: Good point!

File: spec/feature/styleguide_spec.rb (hunk 1 above)
Comment (tsujigiri) [inline spec/feature/styleguide_spec.rb]: It's mentioned in the README.

Comment (dziemba) [inline spec/feature/styleguide_spec.rb]: The README documents why this test file looks strange at the first glance? oO

File: app/cells/hello_world/hello_world_cell.rb (hunk 16 above)
Comment (tsujigiri) [inline app/cells/hello_world/hello_world_cell.rb]: It is still good as an example of a minimal component. I would keep it.

File: config/routes.rb (hunk 11 above)
Comment (dziemba) [inline config/routes.rb]: Okay, but then this file (and many others) are wrong here and belong in the dummy app, right? This exposes all the styleguide displaying logic to the target app.

Comment (tsujigiri) [inline config/routes.rb]: You can still run `rake styleguide` from another app.

Comment (dziemba) [inline config/routes.rb]: That's true, but not my point. This gem is a rails engine, so everything in `config` and `app` will also be autoloaded into the target app. If you are not supposed to run the styleguide directly in the target app, it should not be there, right? `rake styleguide` still works then because it uses the dummy app.

File: lib/ui_components/cell.rb (hunk 12 above)
Comment (dziemba) [inline lib/ui_components/cell.rb]: Okay.

File: config/routes.rb (hunk 11 above)
Comment (tsujigiri) [inline config/routes.rb]: Nobody says you are not supposed to. I could imagine mounting this into the backoffice, so we can look at it without having to start it up locally. I don't see why we shouldn't keep this option.

Comment (dziemba) [inline config/routes.rb]: > So far it is only done in the dummy app and I think that is enough
You just said that...

You want to run this thing on production? The thing that `eval`s code from a YAML file? Seriously?

I think it's best to keep the actual codebase as small as possible and only run this in dev mode. 

One more thing: While starting the dummy app from the target app works for now, it won't work with the released gem: The `spec` directory is usually not packaged into the `.gem` files. So we either have to do that or come up with a better solution.

Comment (tsujigiri) [inline config/routes.rb]: A YAML file we control, yes. How is this different form any other code we put on production?

Good point about the spec dir not being packaged. I didn't know.

Comment (tsujigiri) [inline config/routes.rb]: ...or gems we use, which we don't really have control over.

Comment (tsujigiri) [inline config/routes.rb]: Any objections against adding the dummy app to the files in the gemspec?

Comment (dziemba) [inline config/routes.rb]: You are right, but you also agree with me that putting *any* code that is not necessary for the actual functionality of the app onto production servers is a bad idea, right? Especially since most of the styleguide stuff is barely tested.

Comment (dziemba) [inline config/routes.rb]: No, I think adding the dummy app is okay. Actually we don't even need to do that since we don't release the gem on rubygems. With git references, the whole repo is present anyway.

Comment (tsujigiri) [inline config/routes.rb]: It depends on what you consider the functionality of the app. If you wanted the styleguide mounted in each project, it would kind of belong to the app. If we only want to use it in dev mode, it's fine having it only in the dummy app. I think it would be nice to just have to go to a page instead of having to boot up an app locally, or at least having that option.

Comment (dziemba) [inline config/routes.rb]: Yes, it would also be nice if I didn't have to lock and unlock my apartment door every time I come and go. But I do it anyway, because the small inconvenience definitely outweighs all the downsides of not doing it...

Is typing `rake styleguide` such a big hassle? We can even change the default port to something that's not the rails default.

Comment (tsujigiri) [inline config/routes.rb]: How is code that doesn't even process any input similar to leaving your apartment door open, exactly?

Comment (tsujigiri) [inline config/routes.rb]: Code that doesn't even get executed unless you explicitly mount it, after all.

Comment (dziemba) [inline config/routes.rb]: It still gets loaded... I am not saying this is a security hole, I am just saying it's an unnecessary risk. Also it pollutes the namespace. Now you have a `StyleguideController` class in the global target app namespace. You really want that?

Comment (tsujigiri) [inline config/routes.rb]: OK, OK... I changed it.

File: ui_components.gemspec (hunk 15 above)
Comment (tsujigiri) [inline ui_components.gemspec]: I removed the dummy app's Gemfile.

**Comment by tsujigiri** [comment]
Any last words before I merge this?

**Comment by dziemba** [comment]
So long, and thanks for all the fish.
